qiling/debugger/qdb/qdb.py

```diff
@@ -178,92 +204,74 @@ def do_run(self: QlQdb, *args) -> None:
+        execute one instruction at a time, will enter subroutine
         """
 
-        self.render.context_reg(self._saved_reg_dump)
-        self.render.context_stack()
-        self.render.context_asm()
-
-    def do_backward(self: QlQdb, *args) -> None:
-        """
-        step barkward if it's possible, option rr should be enabled and previous instruction must be executed before
-        """
+        prophecy = self.predictor.predict()
 
-        if self.rr:
-            if len(self.rr.layers) == 0 or not isinstance(self.rr.layers[-1], self.rr.DiffedState):
-                print(f"{color.RED}[!] there is no way back !!!{color.END}")
+        if prophecy.where is True:
+            return True
 
-            else:
-                print(f"{color.CYAN}[+] step backward ~{color.END}")
-                self.rr.restore()
-                self.do_context()
+        if self.ql.archtype == QL_ARCH.CORTEX_M:
+            self.ql.arch.step()
         else:
-            print(f"{color.RED}[!] the option rr not yet been set !!!{color.END}")
+            self._run(count=1)
 
+        self.do_context()
 
-    def update_reg_dump(self: QlQdb) -> None:
+    @SnapshotManager.snapshot
+    @save_reg_dump
+    @check_ql_alive
+    def do_step_over(self: QlQdb, *args) -> Option[bool]:
         """
-        internal function for updating registers dump
+        execute one instruction at a time, but WON't enter subroutine
         """
-        self._saved_reg_dump = dict(filter(lambda d: isinstance(d[0], str), self.ql.reg.save().items()))
 
-    def do_step_in(self: QlQdb, *args) -> Optional[bool]:
-        """
-        execute one instruction at a time, will enter subroutine
-        """
+        prophecy = self.predictor.predict()
 
-        if self.ql is None:
-            print(f"{color.RED}[!] The program is not being run.{color.END}")
+        if prophecy.going:
+            cur_insn = disasm(self.ql, self.cur_addr)
+            self.set_breakpoint(self.cur_addr + cur_insn.size, is_temp=True)
 
         else:
-            self.update_reg_dump()
+            self.set_breakpoint(prophecy.where, is_temp=True)
 
-            prophecy = self.predictor.predict()
+        self._run()
 
-            if prophecy.where is True:
-                return True
+    @SnapshotManager.snapshot
+    @parse_int
+    def do_continue(self: QlQdb, address: Optional[int] = 0) -> None:
+        """
+        continue execution from current address if not specified
+        """
 
-            if self.ql.archtype == QL_ARCH.CORTEX_M:
-                self.ql.arch.step()
-            else:
-                self._run(count=1)
+        if address is None:
+            address = self.cur_addr
 
-                if self.rr:
-                    # context save after execution, so we could do diff
-                    # on two states before and after.
-                    self.rr.save()
+        print(f"{color.CYAN}continued from 0x{address:08x}{color.END}")
```

Comment:
There are a lot of color printing here; consider creating a method to print out colorful messages by specifying the message type (e.g. `qdb_print(msgtype: QDB_MSG, msg: str)` , which may be called: `qdb_print(QDB_MSG.WARNING, "this is a warning!")` )